### PR TITLE
feat: Add Snake Collision method, game mechanics and socket events

### PR DESCRIPTION
**Pull request by vkbinfo**
This implements game logic on the server side By implementing a game class and also added some snake class method for collision.

### CODE REVIEW

File: server/index.js [hunk 1]
```diff
@@ -1,9 +1,13 @@
...
+const cors = require('cors');
+
+const socketIo = require('socket.io');
+const { Game } = require('./Controllers/GameControllers/game');
 
 const PORT = process.env.PORT || 4000;
 
 const app = express();
-
+app.use(cors());
```
Comment (niinpatel) [inline server/index.js]: get the latest changes from master branch, I've already included cors in it.

File: server/app.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+runtime: nodejs
```
Comment (vkbinfo) [inline server/app.yaml]: Remove this file. this is for Google cloud app.

File: server/src/index.js [hunk 3]
```diff
@@ -1,10 +1,38 @@
+import socketIo from 'socket.io';
 import app from './app/app';
-import './database/connection';
+// import './database/connection';
```
Comment (vkbinfo) [inline server/src/index.js]: uncomment this

**Comment by vkbinfo** [comment]
Game logic on the front-end side
snake-game-play-on-front-end

File: server/src/controllers/GameControllers/snake.js [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+    return false;
+  }
+
+  /**
+   * We are drawing arcs on each cooridnate point in the snake body, so to check collision we are getting the most ahead part of the snake head.
+   */
+  // eslint-disable-next-line consistent-return
+  getHeadTip() {
+    const arcRadius = SNAKE_CONSTANTS.ARC_RADIUS;
+    // eslint-disable-next-line default-case
```
Comment (niinpatel) [inline server/src/controllers/GameControllers/snake.js]: No harm in adding a default case here I think, then you don't need eslint-disable for default case and consistent-return

File: server/src/controllers/GameControllers/game.js [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    this.setIntervalNumber(interval);
+  }
+
+  setIntervalNumber(number) {
+    this.firstSnake.interval = number;
+    this.secondSnake.interval = number;
+  }
+
+  makeFood() {
+    // TODO: find place for food on the board by checking that food is not getting created on the snake board;
```
Comment (niinpatel) [inline server/src/controllers/GameControllers/game.js]: Remove all todo comments if they're outdated.

File: server/__tests__/snake.test.js [hunk 6]
```diff
@@ -62,18 +61,135 @@ describe('Test Snake Class', () => {
...
-      expect(snakeLength + SNAKE.FOOD_LENGTH_REWARD).toEqual(snake.bodyCoordinates.length);
-      expect(lifeLine + SNAKE.FOOD_LIFE_REWARD).toEqual(snake.life);
+      expect(snakeLength + SNAKE_CONSTANTS.FOOD_LENGTH_REWARD).toEqual(snake.length);
+      expect(bodyArrayLengthOfSnakeBeforeEating).toEqual(snake.bodyCoordinates.length);
+      expect(lifeLine + SNAKE_CONSTANTS.FOOD_LIFE_REWARD).toEqual(snake.life);
     });
+  });
 
+  describe('method hasCollidedWithItself', () => {
+    it('should return true for this method call after setting given below coordinates', () => {
```
Comment (prakhar1912) [inline server/__tests__/snake.test.js]: Something like `it('checks snake's collisions with itself')` would make more sense than just telling what the function returns.

File: server/__tests__/snake.test.js [hunk 7]
```diff
@@ -62,18 +61,135 @@ describe('Test Snake Class', () => {
...
+        { x: 234, y: 410 }, { x: 234, y: 412 },
+        { x: 234, y: 414 }, { x: 234, y: 416 },
+      ];
+      snake.head = { x: 234, y: 418 };
+      expect(snake.hasCollidedWithItself()).toEqual(true);
+    });
+  });
+
+  describe('method hasCollidedWithRival', () => {
+    it('should return true for this method call after setting given below coordinates', () => {
```
Comment (prakhar1912) [inline server/__tests__/snake.test.js]: Same for this. `it('checks snake's collision with opponent')`

File: server/src/controllers/GameControllers/gameConstants.js [hunk 8]
```diff
@@ -0,0 +1,13 @@
+const SNAKE_CONSTANTS = {
```
Comment (prakhar1912) [inline server/src/controllers/GameControllers/gameConstants.js]: Maybe make this configurable using a `.env`?

File: server/__tests__/snake.test.js (hunk 6 above)
Comment (vkbinfo) [reply server/__tests__/snake.test.js]: Yes, that's sounds good.

File: server/src/controllers/GameControllers/gameConstants.js (hunk 8 above)
Comment (vkbinfo) [reply server/src/controllers/GameControllers/gameConstants.js]: Could it belong in .env file? it is not related to the environment. @niinpatel thoughts?

Comment (niinpatel) [reply server/src/controllers/GameControllers/gameConstants.js]: I missed this! Nope, I don't think this belongs to .env

**Pull request merged**